lib/handler/debug_state.c

```diff
@@ -0,0 +1,236 @@
+            case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+            case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+            case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+            case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+            case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+                return debug_state_string_open;
+            case H2O_HTTP2_STREAM_STATE_IDLE:
+            case H2O_HTTP2_STREAM_STATE_END_STREAM:
+                return NULL;
+        }
+    }
+    return NULL;
+}
+
+static int on_req(h2o_handler_t *_self, h2o_req_t *req)
+{
+    static h2o_generator_t generator = {NULL, NULL};
+
+    /* if the request is sent via HTTP/1, return 404 response */
+    if (req->version < 0x200) {
+        req->res.status = 404;
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("text/plain; charset=utf-8"));
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CACHE_CONTROL, H2O_STRLIT("no-cache, no-store"));
+        h2o_start_response(req, &generator);
+        h2o_send(req, NULL, 0, 1);
+        return 0;
+    }
+
+    h2o_http2_conn_t *conn = (void *)req->conn;
+
+    size_t nr_resp = 4; // others
+    nr_resp += conn->num_streams.pull.open + conn->num_streams.push.open; // streams
+    nr_resp += conn->_input_header_table.num_entries + conn->_output_header_table.num_entries; // hpack
+
+    h2o_iovec_t resp[nr_resp];
+    int cur_resp = 0;
+    memset(resp, 0, sizeof(resp[0]) * nr_resp);
+
+
+#define BUFSIZE (2 * 1024)
+#define OUTPUT(_fmt, ...)                                     \
+    do {                                                      \
+        h2o_iovec_t v;                                        \
+        v.base = h2o_mem_alloc_pool(&req->pool, BUFSIZE);     \
+        v.len = snprintf(v.base, BUFSIZE, _fmt, __VA_ARGS__); \
+        resp[cur_resp++] = v;                                 \
+    } while (0)
+
+#define OUTPUT_HEADER_TABLE(ht)                                                               \
+    do {                                                                                         \
+        h2o_hpack_header_table_t *header_table = ht;                                   \
+        int i, coma_removed = 0;                                                                    \
+        for (i = 0; i < header_table->num_entries; i++) {                                    \
+            h2o_hpack_header_table_entry_t *entry = h2o_hpack_header_table_get(header_table, i); \
+            OUTPUT(                                                                              \
+                   ",\n"                                                                         \
+                   "      [ \"%s\", \"%s\" ]",                                                    \
+                   entry->name->base,                                                            \
+                   entry->value->base);                                                          \
```

Comment:
Please use `%.*s` and specify the lengths of the values using `->len` for both `name` and `value`.

Also, please consider the case in which the name (or value) is very long. Current approach of allocating a fixed amount of memory would not work well.